blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+                    signature,
+                    txs,
+                }))
+            }
+            MessageType::GetBlock => {
+                let height = src.read_u64()?;
+                Ok(Message::GetBlock(GetBlock { height }))
+            }
+            MessageType::Inventory => {
+                let inventory = Inventory::decode(src)?;
+                Ok(Message::Inventory(inventory))
+            }
+            MessageType::GetInventory => {
+                let version = src.read_u64()?;
+                let shortid_nonce = src.read_u64()?;
+                Ok(Message::GetInventory(GetInventory {
+                    version,
+                    shortid_nonce,
+                }))
+            }
+            MessageType::MempoolTxs => {
+                let tip = src.read_blockid()?;
+                let txs = read_block_txs(src)?;
+                Ok(Message::MempoolTxs(MempoolTxs { tip, txs }))
+            }
+            MessageType::GetMempoolTxs => {
+                let shortid_nonce = src.read_u64()?;
+                let shortid_list = src.read_shortid_vec()?;
+                Ok(Message::GetMempoolTxs(GetMempoolTxs {
+                    shortid_nonce,
+                    shortid_list,
+                }))
+            }
+        }
+    }
+}
+
+impl Encodable for Message {
+    fn encode(&self, dst: &mut impl Writer) -> Result<(), WriteError> {
+        match self {
+            Message::Block(b) => {
+                dst.write_u8(b"message type", MessageType::Block as u8)?;
+                BlockHeader::encode(&b.header, dst)?;
+                dst.write_signature(&b.signature)?;
+                write_block_txs(&b.txs, dst)?;
+            }
+            Message::GetBlock(g) => {
+                dst.write_u8(b"message type", MessageType::GetBlock as u8)?;
+                dst.write_u64(b"block height", g.height)?;
+            }
+            Message::Inventory(inv) => {
+                dst.write_u8(b"message type", MessageType::Inventory as u8)?;
+                Inventory::encode(&inv, dst)?;
+            }
+            Message::GetInventory(g) => {
+                dst.write_u8(b"message type", MessageType::GetInventory as u8)?;
+                dst.write_u64(b"version", g.version)?;
+                dst.write_u64(b"shortid_nonce", g.shortid_nonce)?;
+            }
+            Message::MempoolTxs(mempool) => {
+                dst.write_u8(b"message type", MessageType::MempoolTxs as u8)?;
+                dst.write_blockid(b"tip", &mempool.tip)?;
+                write_block_txs(&mempool.txs, dst)?;
+            }
+            Message::GetMempoolTxs(g) => {
+                dst.write_u8(b"message type", MessageType::GetMempoolTxs as u8)?;
+                dst.write_u64(b"shortid_nonce", g.shortid_nonce)?;
+                dst.write_shortid_vec(b"shortid_list", &g.shortid_list)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        unimplemented!()
```

Comment:
one last step for the man, giant leap for the mankind!

Reply:
We will remove this method from Encodable trait (see #437)